### PR TITLE
Initial webservice

### PR DESCRIPTION
**Pull request by bradtaylor**
This is a large PR representing the bulk of the work setting up the service, including the routing, authentication, and notebook conversion. The readme has info on the choice of web framework.

There is still some housekeeping to be done- adding docstrings to the utility methods and creating API documentation.

I also need to create the app configuration for deploying the service on Google App Engine.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,52 @@
...
+Under the hood, Jupyter .ipynb files are JSON documents with a particular [format](https://nbformat.readthedocs.io). Jupyter provides a commandline tool / library [nbconvert](https://nbconvert.readthedocs.io) for interacting with these documents and converting them to various formats, in particular HTML.
+
+This project is essentially nbconvert as a REST service. It takes notebook JSON and returns HTML that can be embedded in your application.
+
+There are two routes:
+
+/status is a health-check endpoint. It accepts GET requests and returns 200 'OK'.
+
+/api/convert is the primary conversion endpoint. It accepts POST requests containing notebook JSON and returns HTML for displaying the notebook's content in the browser.
+This endpoint is authorized via Terra's SAM authorization service. Calls therefore require an auth header supplying a Terra Oauth2 bearer token.
```
Comment (ericsong) [inline README.md]: does it have to be a `Terra` oauth token?

File: main.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+from sanic import Sanic
+from sanic import response
+from utils import perform_notebook_conversion, authorized
+
+
+# Webservice routing
+app = Sanic('douglass')
+app.config.from_pyfile('config.py')
```
Comment (ericsong) [inline main.py]: how is the application going to pick the right config file based on the environment it's running in?

File: main.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+
+@app.get('/')
```
Comment (ericsong) [inline main.py]: `/status` for the status endpoint? not sure what Broad convention is

File: main.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+
+@app.post('/api/convert')
+@authorized(app.config.SAM_ROOT)
+async def convert(request):
+    return response.html(await perform_notebook_conversion(request.json))
+
+
+if __name__ == '__main__':
+    app.run(port=8000)
```
Comment (ericsong) [inline main.py]: does production use a different port?

**Review by ericsong — COMMENTED** [review]
The application code looks good to me. Most of my comments are around how this gets productionalized. Ex. how it handles the different environments.

Also a note that `Terra` is not the only user of this service so the comments/code should not have that assumption baked into them.

File: utils.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                response = await f(request, *args, **kwargs)
+                return response
+            else:
+                raise Forbidden
+        return decorated_function
+    return decorator
+
+
+# Query Terra's authorization service SAM to determine user authorization status. Return auth status boolean
+async def check_sam_authorization(request, sam_root):
```
Comment (ericsong) [inline utils.py]: Also, I believe you can indicate private methods in python by prefixing method names with `__`

File: main.py (hunk 2 above)
Comment (bradtaylor) [reply main.py]: I'm hoping to get some advice on this from the Terra UI team. They wrote a deploy script that copied one of the environment examples from config.ENV.py to config.py I will probably do something similar

File: main.py (hunk 4 above)
Comment (bradtaylor) [reply main.py]: I haven't really touched configuring ports etc for production yet. I'm looking into how Google App Engine deploys work. I'm intending to make an additional PR for production deployments onto GAE

File: utils.py (hunk 5 above)
Comment (bradtaylor) [reply utils.py]: Nice! I never knew that. I'll update this to make use of it

File: main.py (hunk 3 above)
Comment (bradtaylor) [reply main.py]: This should indeed by /status. I will fix

File: README.md (hunk 1 above)
Comment (bradtaylor) [reply README.md]: I'll clarify the comment. It's an Oauth2 token representing the user's Google identity, either from Google SSO or gcloud auth login. That Google identity must be known to Terra (user must be registered and enabled in SAM)

**Comment by bradtaylor** [comment]
@ericsong Thanks for the review! I haven't productionized the deployment yet. My intention is to get some advice from folks on the Terra UI team regarding GAE deploy and then handle that as a follow-on PR.

In speaking of Terra, my intention is to refer to the platform-services layer (the way All of Us talks about 'FireCloud'). Would it be helpful to update this to Terra/FireCloud, in case the latter is how it's referred to in AoU documentation

**Comment by ericsong** [comment]
Oh ok. I didn't realize Terra referred to the entire platform-services layer. In my mind, I was thinking that it was referring to the user facing app like AoU.

In that case, I think the way the term is used makes sense.

**Review by ericsong — APPROVED** [review]

File: utils.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+# Authorization decorator
+# Decorated routes will first check the request to see if the caller is authorized to perform the operation
+# taken from example in Sanic documentation
+# https://sanic.readthedocs.io/en/latest/sanic/decorators.html
+def authorized(sam_root):
+    def decorator(f):
+        @wraps(f)
+        async def decorated_function(request, *args, **kwargs):
+            # Check the Terra authorization service SAM for user auth status
+            user_is_authorized = await __check_sam_authorization(request, sam_root)
```
Comment (calbach) [inline utils.py]: What action is being authorized here? What permission do I need?

Are you just checking that the user is authenticated? (if so, should just rename to is_authenticated)

**Pull request merged**